### ISSUE TITLE
[bug] If a dashboard includes a card someone doesn't have access to, it results in a javascript error rather than an informative message

### ISSUE THREAD

**Issue by exabrial** [metabase/metabase#10247]
First, often in open source, people go unthanked... so thank you to everyone at Metabase for an incredible product and your stewardship of the community around it.

**Describe the bug**
If a dashboard includes a card someone doesn't have access to, it results in a javascript error rather than an informative message

**Logs**
```
[Error] TypeError: undefined is not an object (evaluating 't.type')
getParameterTargetFieldId — app-main.bundle.js:5:278962
(anonymous function) — app-main.bundle.js:5:465491
(anonymous function) — vendor.bundle.js:1:226875
(anonymous function) — vendor.bundle.js:1:226875
(anonymous function) — vendor.bundle.js:1:227543
(anonymous function) — vendor.bundle.js:1:226875
(anonymous function) — app-main.bundle.js:5:1854581
p — vendor.bundle.js:1:2005668
run — vendor.bundle.js:1:1255693
componentWillReceiveProps — vendor.bundle.js:1:1254981
updateComponent — vendor.bundle.js:1:1967094
receiveComponent — vendor.bundle.js:1:1966486
receiveComponent — vendor.bundle.js:1:653066
_updateRenderedComponent — vendor.bundle.js:1:1968325
_performComponentUpdate — vendor.bundle.js:1:1968090
updateComponent — vendor.bundle.js:1:1967367
receiveComponent — vendor.bundle.js:1:1966486
receiveComponent — vendor.bundle.js:1:653066
_updateRenderedComponent — vendor.bundle.js:1:1968325
_performComponentUpdate — vendor.bundle.js:1:1968090
updateComponent — vendor.bundle.js:1:1967367
receiveComponent — vendor.bundle.js:1:1966486
receiveComponent — vendor.bundle.js:1:653066
_updateRenderedComponent — vendor.bundle.js:1:1968325
_performComponentUpdate — vendor.bundle.js:1:1968090
updateComponent — vendor.bundle.js:1:1967367
receiveComponent — vendor.bundle.js:1:1966486
receiveComponent — vendor.bundle.js:1:653066
_updateRenderedComponent — vendor.bundle.js:1:1968325
_performComponentUpdate — vendor.bundle.js:1:1968090
updateComponent — vendor.bundle.js:1:1967367
receiveComponent — vendor.bundle.js:1:1966486
receiveComponent — vendor.bundle.js:1:653066
_updateRenderedComponent — vendor.bundle.js:1:1968325
_performComponentUpdate — vendor.bundle.js:1:1968090
updateComponent — vendor.bundle.js:1:1967367
receiveComponent — vendor.bundle.js:1:1966486
receiveComponent — vendor.bundle.js:1:653066
_updateRenderedComponent — vendor.bundle.js:1:1968325
_performComponentUpdate — vendor.bundle.js:1:1968090
updateComponent — vendor.bundle.js:1:1967367
performUpdateIfNecessary — vendor.bundle.js:1:1966702
performUpdateIfNecessary — vendor.bundle.js:1:653256
b — vendor.bundle.js:1:602595
perform — vendor.bundle.js:1:715835
perform — vendor.bundle.js:1:715835
w — vendor.bundle.js:1:603209
w
closeAll — vendor.bundle.js:1:716423
perform — vendor.bundle.js:1:715922
e — vendor.bundle.js:1:603550
a — vendor.bundle.js:1:784927
enqueueSetState — vendor.bundle.js:1:785936
setState — vendor.bundle.js:1:774208
onStateChange — vendor.bundle.js:1:1256299
[native code]
notify — vendor.bundle.js:1:1997742
notifyNestedSubs — vendor.bundle.js:1:1997376
[native code]
onStateChange — vendor.bundle.js:1:1256325
[native code]
m — vendor.bundle.js:1:1258347
(anonymous function) — app-main.bundle.js:5:82370
w — vendor.bundle.js:1:1882459
(anonymous function) — vendor.bundle.js:1:1882251
r — app-main.bundle.js:5:78818
(anonymous function) — app-main.bundle.js:5:78913
promiseReactionJob

	(anonymous function) (app-main.bundle.js:5:1859018)
	w (vendor.bundle.js:1:1882459)
	(anonymous function) (vendor.bundle.js:1:1882251)
	r (app-main.bundle.js:5:1859220)
	(anonymous function) (app-main.bundle.js:5:1859340)
	promiseReactionJob
```

**To Reproduce**
1. create a card in your personal collection
2. create a dashboard in a public collection and include the card
3. log in as a non-admin
4. visit the dashboard

**Expected behavior**
A clear and concise description of the access problem.

**Information about your Metabase Installation:**
all browsers, Firefox, Chrome, Safari

- Your operating system: (e.x. OS X 10, Windows XP, etc)
OSX

- Your databases: (e.x. MySQL, Postgres, MongoDB, …)
MySql
- Metabase version: (e.x. 0.19.3)
32.9

- Metabase hosting environment: (e.x. Mac app, Elastic Beanstalk, Docker, Heroku, Linux/Ubuntu 12)
ubuntu 18.04, adoptopenjdk 212b04

- Metabase internal database: (e.x. H2 (default), MySQL, Postgres)

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?

I mean on a scale of 1 to 10, with one being the soft whispering of pines on a cool summer night, to 10 being at a pandora concert where the lead singer was replaced by a screaching owl, I'd say this is like a 3/10. It's annoying as hell, but once you figure it out it's easy to fix.

Note: the more honest and specific you are here the more we will take you seriously. 

Duly noted.

**Comment by flamber** [comment]
Hi @exabrial 
It took me a little time to understand - and reproduce your issue, since it's important that the dashboard contains cards from different data sources.
But in that case, I'm sure you're seeing issue #7633

**Comment by paulrosenzweig** [comment]
> with one being the soft whispering of pines on a cool summer night

That sounds wonderful. We need some more 1/10 bugs 😉

@flamber what steps did you take to reproduce? Following the steps in #7633, I don't see the same error, but see a different bug. If I have a dashboard with two questions one of which uses a restricted database, I _can_ see both questions in the dashboard without hitting this error. However, when I click through on the question I get the permissions error.

**Comment by flamber** [comment]
@paulrosenzweig 
- User1 (admin) has access to db1 and db2. User2 is part of group Test and has only access to db1.
- User1 creates a dashboard in Our Analytics, where question1 (user1 PC) is querying db1 and question2 (Our Analytics) is querying db2.
- User2 will then only see a blank dashboard with the text `t is undefined`. ([trace](https://github.com/metabase/metabase/files/3331800/console.log))
- Move both questions to Our Analytics, then it shows the dashboard with both cards and with a console error `fetchData error: You don't have permissions to do that.`, but everything works.
- If I then move question2 to user1 PC, then it shows the dashboard, but only question1, and question2 shows the keys-icon with `Sorry, you don't have permission to see this card.`. Nothing in the console.

**Comment by paulrosenzweig** [comment]
@flamber for some reason I'm still unable to see this. I'm sure I'm doing something wrong. If I swap which question is in my personal collection vs our analytics, both render correctly: 
![image](https://user-images.githubusercontent.com/691495/60214563-be00fa80-9833-11e9-974b-43b94a320eee.png)
![image](https://user-images.githubusercontent.com/691495/60214622-d83ad880-9833-11e9-8070-6a6ffe26c814.png)
(In this view the current user can access the data source for C2 but not C1.)

Are you seeing this on master? Also, do you see this issue when there are no filters on the dashboard? From #7633 it sounded like those were necessary.

**Comment by exabrial** [comment]
We only have one datasource... So it might not be that.

It appeared to be a non-admin accessing a card on a dashboard where the dashboard was in a public collection but the card was created in the admins' private collection

**Comment by flamber** [comment]
@paulrosenzweig Interesting. I cannot reproduce on master, but for the second database, I'm using H2 referencing the Sample Dataset internal file, so not sure if that has something to do with it? On 0.32.9 I was testing a production environment with real databases.
And yes, it only triggers when there's a filter. So that would be #7633 

@exabrial 
I've tried many different ways now, but I cannot reproduce your issue with a single source.
This is tested with clean metadata and Sample Dataset. Tried on 0.32.9 and master.
- User1 in "Administrators", user2 in "All users".
- "All users" does not have database access, but has view right to "Our analytics"
- User1 creates question1 in User1 Personal Collection.
- User1 creates dashboard1 in "Our analytics" and adds question1 as card1.
- User2 login and tries to view dashboard1 - sees dashboard1, but card1 shows "Sorry, you don't have permission to see this card".

**Comment by exabrial** [comment]
Ah yes all these dashboards have multiple sql filters.

**Comment by flamber** [comment]
@exabrial 
That was it! I didn't test with a SQL question. Now I can reproduce, but only when there's a filter.
Pretty sure this is the same problem as #7633
I'm getting a blank dashboard with the text `t is undefined` (Firefox 67):
![image](https://user-images.githubusercontent.com/1447303/60221788-38337e00-9838-11e9-8c5a-feec0e83341c.png)
This is how it looks for a user with DB and card access:
![image](https://user-images.githubusercontent.com/1447303/60221900-a5dfaa00-9838-11e9-98ba-428fec1e8bc6.png)

**Comment by walterl** [comment]
I was able to reproduce @flamber's error... mostly. For me (at 9d88576, on `release-0.33.x`) it says `datasetQuery is undefined` in Firefox, and `Cannot read property 'type' of undefined` in Chromium.

b4d574f fixes the error for me on both browsers.

**Comment by walterl** [comment]
Fixed in #10653.
